ABI break with minor version bump in nlohmann::json

The public API for libxeus includes references to `nlohmann::json` (for instance, https://github.com/jupyter-xeus/xeus/blob/ebd21e9e7cfe143b4d0a6783112cc9006b456915/include/xeus/xdebugger.hpp#L55-L60)

This means that the effective ABI of libxeus is tied to the exact expansion of that template. We encountered this trying to compile a kernel when the local version of `nlohmann::json` was 3.11.3 against a copy of libxeus which had been compiled earlier using `nlohmann::json` 3.11.2.

Attempting to compile xeus-python in these circumstances generates a lot of errors like:

```
main.cpp:(.text+0x58b): undefined reference to `xeus::make_null_debugger(xeus::xcontext&, xeus::xconfiguration const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, nlohmann::json_abi_v3_11_3::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned long, double, std::allocator, nlohmann::json_abi_v3_11_3::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> >, void> const&)'
```

since the expansion of `nlohmann::json` (and hence, the libxeus ABI) has changed from

```
nlohmann::json_abi_v3_11_2::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned
long, double, std::allocator, nlohmann::json_abi_v3_11_2::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> > >
```

to

```
nlohmann::json_abi_v3_11_3::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned
long, double, std::allocator, nlohmann::json_abi_v3_11_3::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> >, void>
```

(extra `, void>` at the end of the template)

I don't have a particularly good idea for how to prevent this short of changing the uses of `nlohmann::json` in the API to serialised JSON strings, which probably adds quite a bit of overhead (in addition to being a breaking change in it's own right).

Reference: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1060164 https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1061007

I've not seen this issue before, so it may be that ABI breaks from nlohmann::json are quite rare.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,7 +75,12 @@ message(STATUS "XEUS_BUILD_TESTS:                ${XEUS_BUILD_TESTS}")
 # Dependencies
 # ============
 
-set(nlohmann_json_REQUIRED_VERSION 3.2.0)
+# nlohmann_json requires libraries that exchange json objects to be linked
+# with the same version of nlohmann_json. Therefore this version should be
+# the same in all xeus components; to do so, downstream projects should not
+# search # directly for nlohmann_json, but rely on find_dependency called by
+# find_package(xeus) instead.
+set(nlohmann_json_REQUIRED_VERSION 3.11)
 set(xtl_REQUIRED_VERSION 0.7)
 
 if (NOT TARGET nlohmann_json)

--- a/xeusConfig.cmake.in
+++ b/xeusConfig.cmake.in
@@ -23,7 +23,9 @@ set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR};${CMAKE_MODULE_PATH}")
 
 include(CMakeFindDependencyMacro)
 find_dependency(xtl @xtl_REQUIRED_VERSION@)
-find_dependency(nlohmann_json @nlohmann_json_REQUIRED_VERSION@)
+# nlohmann_json requires libraries that exchange json objects to be linked
+# with the same version of nlohmann_json. 
+find_dependency(nlohmann_json @nlohmann_json_REQUIRED_VERSION@ EXACT)
 
 # This is required when linking with the static target
 if(NOT EMSCRIPTEN)

